## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [41]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [42]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [43]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [44]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.value_counts("category")

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [45]:
# What three item categories had the most entries?
df.value_counts("category")
print("Using value_counts: consumables, furniture, and software had the most entries.")

Using value_counts: consumables, furniture, and software had the most entries.


In [46]:
consumables_df = df.loc[df["category"] == "consumables"]
consumables_df.value_counts("subcategory")

subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [47]:
# For the category with the most entries, which subcategory had the most entries?
print("Out of all consumables, the subcategory \"bathroom supplies\" had the most entries")

Out of all consumables, the subcategory "bathroom supplies" had the most entries


In [48]:
# Which five clients had the most entries in the data?
client_counts = df.value_counts("client_id")
df_by_client = df.set_index("client_id")

client_counts.head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [49]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = client_counts.index[0:5]


In [50]:
top_client = top_5_clients[0]
top_client_df = df.loc[df["client_id"]== top_client]
total_units = sum(top_client_df["qty"])

In [51]:
# How many total units (the qty column) did the client with the most entries order order?
print(f"The client ordered {total_units} units.")

The client ordered 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [52]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["unit_price"] * df["qty"]

In [53]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_rate(weight):
    return 10 if weight>50 else 7

df["shipping_price"] = df["unit_weight"]*df["qty"] * df["unit_weight"].apply(calculate_shipping_rate)

In [54]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = (df["subtotal"]+df["shipping_price"])*1.0925


In [55]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_cost"] = df["unit_cost"] * df["qty"] +df["shipping_price"]


In [56]:
# Create a column for the profit of each line using line cost and line price

df["profit"] = df["total_price"] - df["total_cost"]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [57]:
# Check your work using the totals above
order1_df = df.loc[df["order_id"] == 2742071]
print(sum(order1_df["total_price"]))

order2_df = df.loc[df["order_id"] == 2173913]
print(sum(order2_df["total_price"]))

order3_df = df.loc[df["order_id"] == 6128929]
print(sum(order3_df["total_price"]))

152560.185075
162210.849375
923292.57305


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [64]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

for client in top_5_clients:
    client_df = df.loc[df["client_id"] == client]
    print(f"Client {client} spent ${sum(client_df["total_price"]):,.2f}")

Client 33615 spent $8,374,039.07
Client 66037 spent $10,256,782.68
Client 46820 spent $9,741,064.88
Client 38378 spent $12,903,598.64
Client 24741 spent $82,265,266.93


In [86]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_data =[]



for client in top_5_clients:
    row = {}
    row["client_id"] = client
    row["total_units"] = sum((df.loc[df["client_id"] == client])["qty"])
    row["total_shipping_price"] = sum(df.loc[df["client_id"] == client]["shipping_price"])
    row["total_revenue"] = sum(df.loc[df["client_id"] == client]["total_price"])
    row["total_profit"] = sum(df.loc[df["client_id"] == client]["profit"])
    summary_data.append(row)

pd.options.display.float_format = '{:.2f}'.format
summary_df = pd.DataFrame(summary_data)
summary_df

client_id  total_units  total_shipping_price  total_revenue  total_profit
0      33615        64313            1825992.21     8374039.07    2201717.84
1      66037        43018            1392651.05    10256782.68    3254800.50
2      46820        75768            1598950.50     9741064.88    2736371.58
3      38378        73667            3426753.12    12903598.64    3271579.94
4      24741       239862            5123130.25    82265266.93   36579685.34

In [87]:
# Format the data and rename the columns to names suitable for presentation.
summary_df = summary_df.rename(columns={"client_id":"Client ID", "total_units":"Total Units", "total_shipping_price":"Total Shipping Price",
                                        "total_revenue":"Total Revenue", "total_profit":"Total Profit"})
# Define the money columns. 

shipping_price = summary_df["Total Shipping Price"]
revenue = summary_df["Total Revenue"]
profit = summary_df["Total Profit"]

# Define a function that converts a dollar amount to millions.
def currecy_format_millions(value):
    return value/1000000

# Apply the currency_format_millions function to only the money columns. 

summary_df["Total Shipping Price"] = shipping_price.apply(currecy_format_millions)
summary_df["Total Revenue"] = revenue.apply(currecy_format_millions)
summary_df["Total Profit"] = profit.apply(currecy_format_millions)
# Rename the columns to reflect the change in the money format. 
summary_df = summary_df.rename(columns={"Total Shipping Price": "Total Shipping Price (millions)",
                                         "Total Revenue": "Total Revenue (millions)", "Total Profit": "Total Profit (millions)"})



Client ID  Total Units  Total Shipping Price (millions)  \
0      33615        64313                             1.83   
1      66037        43018                             1.39   
2      46820        75768                             1.60   
3      38378        73667                             3.43   
4      24741       239862                             5.12   

   Total Revenue (millions)  Total Profit (millions)  
0                      8.37                     2.20  
1                     10.26                     3.25  
2                      9.74                     2.74  
3                     12.90                     3.27  
4                     82.27                    36.58

In [88]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_df = summary_df.sort_values("Total Profit (millions)", ascending=False)

In [89]:
final_df

#Summary in README

Client ID  Total Units  Total Shipping Price (millions)  \
4      24741       239862                             5.12   
3      38378        73667                             3.43   
1      66037        43018                             1.39   
2      46820        75768                             1.60   
0      33615        64313                             1.83   

   Total Revenue (millions)  Total Profit (millions)  
4                     82.27                    36.58  
3                     12.90                     3.27  
1                     10.26                     3.25  
2                      9.74                     2.74  
0                      8.37                     2.20